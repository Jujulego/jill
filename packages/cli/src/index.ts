export { buildCommand } from './commands/build';
export { eachCommand } from './commands/each';
export { infoCommand } from './commands/info';
export { logger, OraLogger } from './logger';
