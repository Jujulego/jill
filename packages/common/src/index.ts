export * from './components';
export * from './filters';
export * from './pipeline';
