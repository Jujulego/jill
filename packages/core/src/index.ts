export * from './git';
export * from './logger';
export * from './project';
export * from './tasks';
export * from './utils';
export * from './workspace';
