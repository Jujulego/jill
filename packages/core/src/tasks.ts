import { TaskManager } from '@jujulego/tasks';

import { logger } from './logger';

// Setup global manager
export const manager = new TaskManager<any>({ logger });
