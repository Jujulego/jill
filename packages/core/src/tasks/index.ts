export * from './task';
export * from './spawn';