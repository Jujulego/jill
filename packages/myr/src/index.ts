import { Plugin } from '@jujulego/jill-common';

import { MyrCommand } from './myr.command';
import { WatchCommand } from './watch.command';

// Plugin
const myrPlugin = Plugin.createPlugin('myr', [
  MyrCommand,
  WatchCommand
]);

export default myrPlugin;
export { WatchTaskFragment } from './common/watch-task.model';
