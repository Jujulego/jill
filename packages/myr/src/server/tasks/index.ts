export * from './spawn-task.args';
export { TaskIDArgs } from './task-id.args';
