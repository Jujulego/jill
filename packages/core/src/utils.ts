import { streamEvents } from '@jujulego/event-tree';
import { SpawnTask, SpawnTaskStream } from '@jujulego/tasks';

// Utils
export async function* combine<T>(...generators: AsyncGenerator<T>[]): AsyncGenerator<T> {
  for (const gen of generators) {
    yield* gen;
  }
}

export async function *streamLines(task: SpawnTask, stream: SpawnTaskStream): AsyncGenerator<string> {
  // Abort
  const ctrl = new AbortController();

  task.subscribe('status.done', () => ctrl.abort());
  task.subscribe('status.failed', () => ctrl.abort());

  // Stream
  let current = '';

  try {
    for await (const chunk of streamEvents(task, `stream.${stream}`, { signal: ctrl.signal })) {
      const data = current + chunk.data.toString('utf-8');
      const lines = data.split(/\r?\n/g);

      current = lines.pop() ?? '';

      for (const line of lines) {
        yield line;
      }
    }
  } catch (err) {
    if (task.exitCode !== 0) {
      throw err;
    }

    if (current) {
      yield current;
    }
  }
}
