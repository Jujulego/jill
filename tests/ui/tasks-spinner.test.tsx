import { SpawnTask, TaskManager } from '@jujulego/tasks';
import { render, cleanup } from 'ink-testing-library';

import { TasksSpinner } from '../../src/ui';
import { spyLogger } from '../utils';

// Setup
let taskA: SpawnTask;
let taskB: SpawnTask;
let taskC: SpawnTask;

beforeEach(() => {
  taskA = new SpawnTask('cmd-a', [], {}, { logger: spyLogger });
  taskB = new SpawnTask('cmd-b', [], {}, { logger: spyLogger });
  taskC = new SpawnTask('cmd-c', [], {}, { logger: spyLogger });
});

afterEach(() => {
  cleanup();
});

// Tests
describe('<TasksSpinner>', () => {
  it('should print every managed tasks', () => {
    const manager = new TaskManager({ jobs: 0, logger: spyLogger });
    manager.add(taskA);
    manager.add(taskB);

    const { lastFrame } = render(<TasksSpinner manager={manager} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('should print added tasks while running', async () => {
    const manager = new TaskManager({ jobs: 0, logger: spyLogger });
    manager.add(taskA);
    manager.add(taskB);

    const { lastFrame } = render(<TasksSpinner manager={manager} />);

    manager.add(taskC);
    await new Promise((resolve) => setTimeout(resolve, 0));

    expect(lastFrame()).toMatchSnapshot();
  });
});
