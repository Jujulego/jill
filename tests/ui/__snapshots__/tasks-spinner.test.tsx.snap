// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TasksSpinner> should print added tasks while running 1`] = `
"⠋ cmd-a
⠋ cmd-b
⠋ cmd-c"
`;

exports[`<TasksSpinner> should print every managed tasks 1`] = `
"⠋ cmd-a
⠋ cmd-b"
`;
